src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java

```diff
@@ -1264,7 +1267,7 @@ public Set<KeyStore.Entry.Attribute> engineGetAttributes(String alias) {
             return super.engineGetAttributes(alias);
         }
         Entry entry = entries.get(alias.toLowerCase(Locale.ENGLISH));
-        return Collections.unmodifiableSet(new HashSet<>(getAttributes(entry)));
+        return Collections.unmodifiableSet(new HashSet<>(entry.attributes));
```

Comment:
I do not think this fix is correct. A call to `engineGetAttribute()` could observe an empty HashMap set by `populateEntry`, but before the entry is populated. If we want to prevent that I believe we need a lock of some kind shared between `engineGetAttribute` and `populateEntry`. 
Also the constructor of HashSet that takes a collection calls addAll(), which loops over the provided collection. In a multi threaded context, if populateEntry can run concurrently with engineGetAttribute, it is not impossible that this loop will get a ConcurrentModificationException, like before. Or am I missing something?

Reply:
Oh - I see - I did miss that. LGTM then.